feat: Add `dt.replace`

mcrumiller opened the pull request:
Closes #8879.

@MarcoGorelli since this involves constructing new `Date`/`Datetime` chunked arrays, I moved some of the `fn datetime()` construction code that needed to be reused into `DateChunked::new_from_parts` and` DatetimeChunked::new_from_parts`.

### Examples

```python
import polars as pl
from polars import col
from datetime import date, datetime

# -- Series
s = pl.Series([date(2024, 1, 1), date(2024, 1, 2), date(2024, 1, 3)])
s.dt.replace(month=2, day=15)
# shape: (3,)
# Series: '' [date]
# [
#         2024-02-15
#         2024-02-15
#         2024-02-15
# ]

# -- Expr
df = pl.DataFrame({
    "a": [datetime(1, 1, 1, 1, 1, 1, 1)],
})
df.with_columns(
    col("a").dt.replace(year=2024).alias("year"),
    col("a").dt.replace(year=2015, day=23, second=45, microsecond=5000).alias("many"),
)
# shape: (1, 3)
# ┌────────────────────────────┬────────────────────────────┬─────────────────────────┐
# │ a                          ┆ year                       ┆ many                    │
# │ ---                        ┆ ---                        ┆ ---                     │
# │ datetime[μs]               ┆ datetime[μs]               ┆ datetime[μs]            │
# ╞════════════════════════════╪════════════════════════════╪═════════════════════════╡
# │ 0001-01-01 01:01:01.000001 ┆ 2024-01-01 01:01:01.000001 ┆ 2015-01-23 01:01:45.005 │
# └────────────────────────────┴────────────────────────────┴─────────────────────────┘

# -- Expr using other columns
df = pl.DataFrame({
    "date": [date(2024, 1, 1), date(2024, 1, 2)],
    "day": [6, 13],
})
df.with_columns(
    col("date").dt.replace(day="day")
)
# shape: (2, 2)
# ┌────────────┬─────┐
# │ date       ┆ day │
# │ ---        ┆ --- │
# │ date       ┆ i64 │
# ╞════════════╪═════╡
# │ 2024-01-06 ┆ 6   │
# │ 2024-01-13 ┆ 13  │
# └────────────┴─────┘
```

mcrumiller commented on crates/polars-plan/src/dsl/function_expr/temporal.rs:
```diff
@@ -95,91 +95,60 @@ pub(super) fn datetime(
     }
-    let _ambiguous = _ambiguous.str()?;
-
-    let ca: Int64Chunked = year
```
I moved this code into `datetime_series_from_parts` in polars_time/src/series/mod.rs for re-use.

mcrumiller commented on crates/polars-plan/src/dsl/function_expr/temporal.rs:
```diff
@@ -95,91 +95,60 @@ pub(super) fn datetime(
-    let ca: Int64Chunked = year
-        .into_iter()
-        .zip(month)
-        .zip(day)
-        .zip(hour)
-        .zip(minute)
-        .zip(second)
-        .zip(microsecond)
-        .map(|((((((y, m), d), h), mnt), s), us)| {
-            if let (Some(y), Some(m), Some(d), Some(h), Some(mnt), Some(s), Some(us)) =
-                (y, m, d, h, mnt, s, us)
-            {
-                NaiveDate::from_ymd_opt(y, m, d)
-                    .and_then(|nd| nd.and_hms_micro_opt(h, mnt, s, us))
-                    .map(|ndt| match time_unit {
-                        TimeUnit::Milliseconds => ndt.and_utc().timestamp_millis(),
-                        TimeUnit::Microseconds => ndt.and_utc().timestamp_micros(),
-                        TimeUnit::Nanoseconds => ndt.and_utc().timestamp_nanos_opt().unwrap(),
-                    })
-            } else {
-                None
-            }
-        })
-        .collect_trusted();
-
-    let ca = match time_zone {
-        #[cfg(feature = "timezones")]
-        Some(_) => {
-            let mut ca = ca.into_datetime(*time_unit, None);
-            ca = replace_time_zone(&ca, time_zone, _ambiguous, NonExistent::Raise)?;
-            ca
-        },
-        _ => {
-            polars_ensure!(
-                time_zone.is_none(),
-                ComputeError: "cannot make use of the `time_zone` argument without the 'timezones' feature enabled."
-            );
-            ca.into_datetime(*time_unit, None)
-        },
-    };
```
I moved this code into datetime_series_from_parts in polars_time/src/series/mod.rs for re-use.

mcrumiller commented:
@MarcoGorelli I could also perhaps use some help creating tests to cover the `non_existent` parameter and the `ambiguous` parameter, neither of which my tests cover.

MarcoGorelli commented:
> @MarcoGorelli I could also perhaps use some help creating tests to cover the `non_existent` parameter and the `ambiguous` parameter, neither of which my tests cover.

sure - would it work to start with
```
datetime(2020, 10, 25)
```
and then do `.dt.replace(hour=1)`?

mcrumiller commented:
> sure - would it work to start with
> 
> ```
> datetime(2020, 10, 25)
> ```
> 
> and then do `.dt.replace(hour=1)`?

Thanks--looks like I need to use `Europe/London` for this to work (found in another unit test).

eitsupi commented on py-polars/polars/expr/datetime.py:
```diff
@@ -390,6 +390,40 @@ def round(self, every: str | dt.timedelta | IntoExprColumn) -> Expr:
 
+    def replace(
+        self,
+        year: int | IntoExpr | None = None,
+        month: int | IntoExpr | None = None,
+        day: int | IntoExpr | None = None,
+        hour: int | IntoExpr | None = None,
+        minute: int | IntoExpr | None = None,
+        second: int | IntoExpr | None = None,
+        microsecond: int | IntoExpr | None = None,
+        ambiguous: Ambiguous | Expr = "raise",
+        non_existent: NonExistent = "raise",
```
I suppose they should be keyword-only arguments.

MarcoGorelli commented on crates/polars-plan/src/dsl/function_expr/temporal.rs:
```diff
@@ -95,91 +94,61 @@ pub(super) fn datetime(
     let year = year.i32()?;
 
-    let mut month = month.cast(&DataType::UInt32)?;
+    let mut month = month.cast(&DataType::Int8)?;
```
why does this need to change?

mcrumiller replied:
Hi Marco--it doesn't, but I was trying to be consistent elsewhere where for e.g. `ca.month()` returns a `i8`, so now that we have `new_from_parts(...)` it made sense to me that the `month` input should be a `i8` as well.

I understand that it pretty much immediately gets cast back to an `i32` so I'll revert this minor change.

mcrumiller replied:
@MarcoGorelli what's your opinion here? Should I revert and keep it to u32?

MarcoGorelli reviewed (commented):
thanks for working on this, it'll probably be useful to people

eitsupi commented on py-polars/polars/expr/datetime.py:
```diff
@@ -390,6 +390,38 @@ def round(self, every: str | dt.timedelta | IntoExprColumn) -> Expr:
 
+    def replace(
+        self,
+        year: int | IntoExpr | None = None,
+        month: int | IntoExpr | None = None,
+        day: int | IntoExpr | None = None,
+        hour: int | IntoExpr | None = None,
+        minute: int | IntoExpr | None = None,
+        second: int | IntoExpr | None = None,
+        microsecond: int | IntoExpr | None = None,
+        ambiguous: Ambiguous | Expr = "raise",
```
I suppose they should be keyword-only arguments.

mcrumiller replied:
They are already, no?

mcrumiller replied:
Done, thanks for the suggestion @eitsupi.

eitsupi replied:
Thanks!

ritchie46 commented on crates/polars-time/src/replace.rs:
```diff
@@ -0,0 +1,187 @@
+        // 3. Value was supplied and is Series   --> Update all elements with the non-null values
+        let year = if year.len() == 1 {
+            // SAFETY: array has one value.
+            if let Some(value) = unsafe { year.get_unchecked(0) } {
```
No need to use unsafe for a single value.

mcrumiller replied:
Fixed in [34d39ea](https://github.com/pola-rs/polars/pull/19708/commits/34d39ea3b0613036bd988dbb5ce9d6cbab2dcd67). I'll resolve all of these.

ritchie46 commented on crates/polars-time/src/replace.rs:
```diff
@@ -0,0 +1,187 @@
+        };
+        let month = if month.len() == 1 {
+            // SAFETY: array has one value.
+            if let Some(value) = unsafe { month.get_unchecked(0) } {
```
No need to use unsafe for a single value.

ritchie46 commented on crates/polars-time/src/replace.rs:
```diff
@@ -0,0 +1,187 @@
+        };
+        let hour = if hour.len() == 1 {
+            // SAFETY: array has one value.
+            if let Some(value) = unsafe { hour.get_unchecked(0) } {
```
No need to use unsafe for a single value.

ritchie46 commented on crates/polars-time/src/replace.rs:
```diff
@@ -0,0 +1,187 @@
+        };
+        let day = if day.len() == 1 {
+            // SAFETY: array has one value.
+            if let Some(value) = unsafe { day.get_unchecked(0) } {
```
No need to use unsafe for a single value.

ritchie46 commented on crates/polars-time/src/replace.rs:
```diff
@@ -0,0 +1,187 @@
+        };
+        let microsecond = if microsecond.len() == 1 {
+            // SAFETY: array has one value.
+            if let Some(value) = unsafe { microsecond.get_unchecked(0) } {
```
No need to use unsafe for a single value.

ritchie46 commented on crates/polars-time/src/replace.rs:
```diff
@@ -0,0 +1,187 @@
+        };
+        let second = if second.len() == 1 {
+            // SAFETY: array has one value.
+            if let Some(value) = unsafe { second.get_unchecked(0) } {
```
No need to use unsafe for a single value.

ritchie46 commented on crates/polars-time/src/replace.rs:
```diff
@@ -0,0 +1,187 @@
+        };
+        let minute = if minute.len() == 1 {
+            // SAFETY: array has one value.
+            if let Some(value) = unsafe { minute.get_unchecked(0) } {
```
No need to use unsafe for a single value.

ritchie46 commented on crates/polars-time/src/replace.rs:
```diff
@@ -0,0 +1,187 @@
+        };
+        let day = if day.len() == 1 {
+            // SAFETY: array has one value.
+            if let Some(value) = unsafe { day.get_unchecked(0) } {
```
No need to use unsafe for a single value.

ritchie46 commented on crates/polars-time/src/replace.rs:
```diff
@@ -0,0 +1,187 @@
+        } else {
+            &year.zip_with(&year.is_not_null(), &self.year())?
+        };
+        let month = if month.len() == 1 {
```
No need to use unsafe for a single value.

ritchie46 commented on crates/polars-time/src/replace.rs:
```diff
@@ -0,0 +1,187 @@
+
+        let year = if year.len() == 1 {
+            // SAFETY: array has one value.
+            if let Some(value) = unsafe { year.get_unchecked(0) } {
```
No need to use unsafe for a single value.

mcrumiller commented on crates/polars-time/src/replace.rs:
```diff
@@ -0,0 +1,187 @@
+use crate::prelude::*;
+
+#[cfg(feature = "dtype-datetime")]
+pub trait PolarsReplaceDatetime: DatetimeMethods {
```
I think fixed in [841dfb6](https://github.com/pola-rs/polars/pull/19708/commits/841dfb641615bc00e65c515d7be1cb0915077318). Is this what you intended?

ritchie46 replied:
Yes!

ritchie46 commented on crates/polars-time/src/replace.rs:
```diff
@@ -0,0 +1,155 @@
+        &month.zip_with(&month.is_not_null(), &ca.month())?
+    };
+    let day = if day.len() == 1 {
+        // SAFETY: array has one value.
```
Stray safety comments

ritchie46 commented on crates/polars-time/src/replace.rs:
```diff
@@ -0,0 +1,155 @@
+    } else {
+        &year.zip_with(&year.is_not_null(), &ca.year())?
+    };
+    let month = if month.len() == 1 {
```
This file has got stray safety comments.

mcrumiller replied:
Fixed.

mcrumiller commented:
@ritchie46 All of your comments have been addressed, I think this is good to go?

MarcoGorelli commented on py-polars/polars/expr/datetime.py:
```diff
@@ -390,6 +390,98 @@ def round(self, every: str | dt.timedelta | IntoExprColumn) -> Expr:
+        ...     }
+        ... )
+        >>> df.with_columns(
+        ...     pl.col("date").dt.replace(
+        ...         year="year", month="month", day="day", hour="hour", minute="minute"
+        ...     )
```
i'd suggest keeping the example simpler - just replace 1 or 2 fields, but most importantly, alias the result, so that it shows before vs after

MarcoGorelli commented on py-polars/polars/series/datetime.py:
```diff
@@ -2096,3 +2096,61 @@ def dst_offset(self) -> Series:
+        shape: (2,)
+        Series: 'datetime' [datetime[μs]]
+        [
+                2022-01-04 12:15:00
+                2022-01-04 12:15:00
```
the result is the same for both rows, could we have a better example which shows different ones? like just `s.dt.replace(year=2022)`?

mcrumiller replied:
Yep, good idea, bad on my part.

MarcoGorelli reviewed (commented):
thanks, just left some comments

the coverage report looks a little concerning

> Attention: Patch coverage is 81.72043% with 51 lines in your changes missing coverage. Please review.

have you checked whether this is a false alarm, or whether those 51 lines are really uncovered? if the latter, then tbh I think they really need covering

mcrumiller commented on crates/polars-python/src/expr/datetime.rs:
```diff
@@ -99,6 +99,33 @@ impl PyExpr {
 
+    fn dt_replace(
+        &self,
+        day: Self,
+        month: Self,
+        year: Self,
+        hour: Self,
+        minute: Self,
+        second: Self,
+        microsecond: Self,
```
Mistake, good catch.

mcrumiller commented:
> the coverage report looks a little concerning

You're right, looks like we're not hitting the `Date` path in the tests, only `Datetime`. Will fix, thanks.

mcrumiller commented:
@MarcoGorelli something new I encountered, I'm not sure what expected behavior should be: if we have a series with null values, and we provide a value for all time components in `replace`, should the null propagate? Example:

```
s = pl.Series([None], dtype=pl.Date)
s.dt.replace(year=2000, month=1, day=1)
```
Should the result be the date 2000-01-01, or should it remain null?

adamreeve commented:
> Should the result be the date 2000-01-01, or should it remain null?

In my opinion, it should be null. Making a non-null result would be inconsistent and quite surprising.

MarcoGorelli commented:
yup, agree with propagating nulls

mcrumiller commented:
@MarcoGorelli I believe I have addressed all of your issues. In addition, nulls will now propagate even if all time components are provided (this is included in the tests).

ritchie46 commented:
Alright, I will leave last decision here to @MarcoGorelli

mcrumiller commented on crates/polars-time/src/replace.rs:
```diff
@@ -0,0 +1,163 @@
+        let mask = &ca.is_not_null();
+        let null = &Int32Chunked::full_null(PlSmallStr::EMPTY, n);
+        out = out.physical().zip_with(mask, null)?.into_date();
+    }
```
@ritchie is there a more efficient way to zip validities here?

ritchie46 replied:
Yes, use arrow arrays. They have bitand/or kernels.

mcrumiller replied:
@MarcoGorelli I'm going to re-draft this briefly while I figure out how to do this part a bit more efficiently.

MarcoGorelli commented on crates/polars-time/src/replace.rs:
```diff
@@ -0,0 +1,156 @@
+        if let Some(value) = microsecond.get(0) {
+            &Int32Chunked::full("".into(), value, n)
+        } else {
+            &(ca.nanosecond() / 1000)
```
are you sure this is right?

mcrumiller replied:
@MarcoGorelli you're right, I should look into using `and_hms_nano_opt` for the datetime construction.

mcrumiller replied:
How do we want to do this? Have `microseconds` and `nanoseconds` as mutually exclusive input arguments? Or perhaps `subseconds` which is dependent on the time unit? That's a bit less obvious to the end-user that doesn't supply a time unit in the construction.

MarcoGorelli replied:
I think the original nanoseconds need to be preserved, there doesn't necessarily need to be a `nanoseconds` argument (at least, not in this PR, it could be a separate topic for add one here and in `datetime`)

MarcoGorelli replied:
As in, the user-facing API could be kept as-is, but `DatetimeChunked::new_from_parts` takes `nanosecond` as argument instead of `microsecond`

then here you make a `nanosecond` variable and pass that

mcrumiller replied:
@MarcoGorelli the only issue I'm running into is with nanosecond Series, trying to replace values that are outside of the allowable range. Should I just let the panic happen?

This issue exists elsewhere in the codeback, e.g. here:

```python
from datetime import datetime
pl.Series(datetime(1500, 1, 1), dtype=pl.Datetime("ns"))
shape: (1,)
Series: '' [datetime[ns]]
[
        2084-07-20 23:34:33.709551616
]
```
Note the wrong datetime. Much of the construction that goes through chrono silently fails with outside-of-bounds dates. In the case now, if we try to replace outside of the allowable range, it'll panic:

```python
s = pl.Series([datetime(2000, 1, 1)], dtype=pl.Datetime("ns"))
s.dt.replace(year=1900)  # ok, works
s.dt.replace(year=1500)  # panic
```
Not sure whether we should just allow this as a limitation. Catching the error and raising may slow things down a bit.

MarcoGorelli replied:
that's fine, that should error (there's a separate issue about making it error instead of panic), if someone has `'ns'` time unit they need to stay within the bounds of that time unit

mcrumiller replied:
@MarcoGorelli I think in that case the new version is ready for review. I've added in two additional tests to ensure that 1) nanoseconds are preserved if microseconds are not provided, and 2) that time zones & time units are preserved.

MarcoGorelli replied:
thanks - could you rebase or merge please?

mcrumiller replied:
@MarcoGorelli done.

MarcoGorelli reviewed (changes requested):
thanks - the microsecond / nanosecond part looks a bit off?
```python
In [12]: df = pl.DataFrame({'a': ['2020-01-01T00:00:00.123456789']*2}).select(pl.col('a').str.to_dat
    ...: etime(time_unit='ns'))

In [13]: df
Out[13]: 
shape: (2, 1)
┌───────────────────────────────┐
│ a                             │
│ ---                           │
│ datetime[ns]                  │
╞═══════════════════════════════╡
│ 2020-01-01 00:00:00.123456789 │
│ 2020-01-01 00:00:00.123456789 │
└───────────────────────────────┘

In [14]: df['a'].dt.replace(year=2021)
Out[14]: 
shape: (2,)
Series: 'a' [datetime[ns]]
[
        2021-01-01 00:00:00.123456
        2021-01-01 00:00:00.123456
]
```

MarcoGorelli commented on crates/polars-time/src/replace.rs:
```diff
@@ -0,0 +1,157 @@
+        &nanosecond.zip_with(&nanosecond.is_not_null(), &ca.nanosecond())?
+    };
+
+    println!("nanosecond: {:?}", nanosecond);
```
mate

mcrumiller replied:
Yikes sorry, fixed.

MarcoGorelli reviewed (commented):
print statement

MarcoGorelli commented:
😄 no worries, thanks for updating

looks like there's another merge conflict, sorry

mcrumiller commented:
> looks like there's another merge conflict, sorry

Figured that would happen that with that PR. Updated.

MarcoGorelli reviewed (approved):
Nice one, thanks @mcrumiller !

Leaving open a bit in case there's objections

separately to this PR, as a follow-up, `py-polars/tests/unit/operations/namespaces/temporal/test_datetime.py` has become very big, I think it'd be worth splitting these `dt.replace` tests into a separate test file
